### DISCUSSION TITLE
Moodle Doesn't Accept SEB Server-Generated .seb Files

### DISCUSSION THREAD

**Discussion by blanderson22** [SafeExamBrowser/seb-server#70] [Q&A]
When I configure a Moodle quiz to require the use of Safe Exam Browser to attempt the quiz, I then choose to upload a config file, I select the file I generated via SEB Server. Moodle then prevents the saving of the quiz settings at that point with the following error: 

` The uploaded file could not be saved as a SEB config file. `

Based on the Moodle forums, this has been an issue since around 2020/2021. Are there certain settings I need to set or remove within SEB Server in order for the config file to be Moodle-friendly?

**Comment by blanderson22** [comment]
I attempted to open a .seb config file locally with a text editor to compare the contents to a .seb config file downloaded directly from my Moodle instance. While the Moodle .seb opened in a text editor, I attempted to do the same with the SEB Server .seb file and I received an encoding error and the file would not open. So, perhaps the encoding issue is the same reason Moodle will not accept the SEB Server .seb file.

**Comment by blanderson22** [comment]
It seems that it depends on where within SEB Server I export the .seb file as to whether it's accepted by Moodle. For instance, exporting from Exam > Export Exam Connection Configuration results in a file that Moodle will not accept. 
However, exporting from Exam Configuration > View SEB Settings results in a file that Moodle does accept.

**Reply by anhefti** [reply]
Hi,

Yes, as described in the user-manual, in SEB SEB Server there are two different kind of SEB configurations:

- **Connection Configuration** : This configuration is to configure SEB to securely connect to the SEB Server for applying an exam setup with SEB Server. This can only be used with SEB and SEB Server. 
You can download a Connection Configuration that is not specific for an exam from the Connection Configuration section. Or if you have setup an exam, you also can download a Connection Configuration from the Exam page "Export Exam Connection Configuration" and choose a defined  Connection Configuration in the second step. With the first download you get a SEB configuration that connect to SEB Server and SEB will present you a list of running exams that are defined within the SEB Server. With the second described download you get a Connection Configuration that is specific for an exam, SEB will connect to SEB Server and automatically select/connect to the appropriate exam on SEB Server.

- **Exam Configuration**: This is the usual SEB configuration for an exam and is usually not needed to be downloaded when an exam is been done with SEB Server. SEB will connect to the SEB Server with the downloaded Connection Configuration and at the time the participating exam is clear, SEB will automatically download the Exam Configuration from SEB Server for that particular exam and after that SEB will connect to the LMS/Moodle.  You can download the plain XML file of such a Exam Configuration but this is usually only for testing.

At the moment SEB Server  is not fully compatible with the SEB deeper integration Plugin of Moodle. But we are working on a SEB Server Plugin for Moodle to allow automated SEB exam restrictions on Moodle site too.

**Comment by blanderson22** [comment]
Hi Andreas -

Yes, I understand the difference in the two files' purpose. I was just
posting a comparison in troubleshooting the issue I originally posted.
There seems to be an encoding issue with the Exam-specific .seb file. I
received an error when I tried to view its contents in a text editor, and
also, Moodle reports it as an invalid .seb file.


Thanks,
Bridgette

--------------------------
Bridgette Anderson
3by400.com
678-379-9321 :: 706-463-1003

**Reply by anhefti** [reply]
Hi, Yes, the download from the Exam with "Export Exam Connection Configuration" is a **Connection Configuration** (specific for this exam).
A Connection Configuration is zipped and encrypted if you have selected encryption (either with password or certificate). If you have just downloaded it (with no encryption), it is at least zipped and you should be able to open it with a zip tool. I expect that Moodle cannot handle zipped or encrypted .seb files.

**Comment by blanderson22** [comment]
I downloaded and extracted the file but attempting to open the contents for
reading still generates an encoding error. The contents are a single
document with no file ext. I tried opening it for reading without an
extension, and I also tried adding a .seb extension and opening it ... the
same end result.

Thanks,
Bridgette

--------------------------
Bridgette Anderson
3by400.com
678-379-9321 :: 706-463-1003

**Reply by anhefti** [reply]
Can you post this file here, then we can have a look at it. Thanks

**Reply by blanderson22** [reply]
[pmve-mock BeginMockTestSession_2022-10-6.seb.zip](https://github.com/SafeExamBrowser/seb-server/files/9727792/pmve-mock.BeginMockTestSession_2022-10-6.seb.zip)

**Comment by anhefti** [comment]
Thanks,

Hmm it seems that you have zipped it yourself on your Mac?
I assume  that you have done this because .seb files are not accepted here right?

Ah I see, the file is doubled zipped because of the SEB file format described [here](https://safeexambrowser.org/developer/seb-file-format.html)

The unzipped file looks something like this in a texteditor:

![grafik](https://user-images.githubusercontent.com/39078777/194486463-648a1716-2701-4639-a630-287f30f6649e.png)

The "plnd" prefix indicates that it is not encrypted, just zipped again. So if you want you can open this file again within a zip-tool and will get a plain text XML SEB settings file:

![grafik](https://user-images.githubusercontent.com/39078777/194486996-d4dc8b98-8065-44d0-8317-5aa8dd3aafe8.png)


But at least Moodle should be able to handle the extracted file with the "plnd" prefix. This might be an issue on Moodle side. You should be able to load the file (the original as well as the unzipped) file into your SEB Configuration Tool (Windows) or with SEB itself. So it is a valid SEB file.

May I ask you what your use-case is here? Because a SEB Server Connection Configuration was anyway never meant to be used on Moodle. If you want to setup an exam with SEB Server and Moodle deeper SEB integration, to restrict your exam on Moodle for SEB access only, there is a workaround described [here]( https://github.com/SafeExamBrowser/seb-server/issues/25#issuecomment-825580323) for that.
As mentioned, the two, SEB Server and Moodle deeper SEB integration are currently not really compatible but we are working on a Moodle plugin for SEB Server that will address this feature. 

Best
Andreas

**Reply by danschlet** [reply]
Please note that .seb files use gzip as compression algorithm, not zip itself.

**Reply by blanderson22** [reply]
I guess I'm a bit confused here, because we also have Zoom integration setup for live proctoring on our SEB Server, and the only location where the SEB Server requests Live Proctoring Zoom App/SDK keys is in the Exam Administration > Exam. This is also where I retrieved the exported config from, and I did so because 1) it's specific to the exam I'm working with in Moodle, and, 2) it contains the Zoom connection settings. So, my thinking was that is the seb config file I must provide to students in order for them to be connected to the correct Exam and Zoom meeting via SEB Server. 

But, I think what you're saying, based on the documentation (which maybe I didn't fully understand heh) is that I really should be downloading the Configurations > Connection Configuration export, and then when students launch that config file the SEB client will then download the correct exam configuration from the SEB Server. And, that exam config file will connect the student to the correct exam and zoom meeting?

**Reply by anhefti** [reply]
Yes this is partly correct. 

You can use Configurations > Connection Configuration export as well as Exam Administration > Exam > Export Exam Connection Configuration (red in the below picture) and then launch that config file with SEB client. Since both of them are connection configurations.
In the first case the connection configuration is not specific to an exam and SEB connects to the SEB Server and SEB Server sends a list of all running exams to SEB. SEB shows this to the students for selection. Student can select one and then SEB download the Exam Configuration (blue) from the SEB Server (for the selected exam).
In the second case the Connection Configuration is specific to the exam and the above step is skipped. SEB client connects to the SEB Server and directly download the Exam Configuration that has been defined or this exam (blue in the below picture) 

Please note the the above files do not have to be uploaded to Moodle and SEB shall not connect to Moodle first when you use SEB Server. Instead the above files shall be started/lunched directly with SEB. Of course you can put the file as a downloadable .seb file link to any kind of Portal where SEB is connecting to or send it by Mail to the students to lunch it with SEB. The important point here is that SEB has fist to connect to SEB Server not to Moodle. 

![image](https://user-images.githubusercontent.com/39078777/196108282-2a5be7e5-8a72-4612-8bf2-9a407ed4268b.png)

This is independent from the kind of LMS you use and just the usual workflow for SEB with SEB Server integration. 
The whole setup is a bit more complex when it came to Moodle and SEB deeper integration plugin. 
If you don't want to restrict the exam on Moodle with the ConfigKey you can just disable the SEB deeper integration on Moodle side and use the setup as described above. Use the selection "No" on Moodle side for the "Required the use of Safe Exam Browser".

If you want to restrict the exam on Moodle with the ConfigKey, you have to put the settings in the Exam Configuration (blue) as it is described within the workaround, download the SEB Settings from the Exam Configuration and upload that to Moodle, if I remember this correctly, use the selection "Yes - Use my own config" on Moodle

**Comment by blanderson22** [comment]
Yes, I zipped it in order to upload it to the discussion. The .seb file
wasn’t accepted.

Regarding my inability to open the file for reading, perhaps my unarchiver
app unzipped it rather than ungzipping it ;)

Yes, I did previously find the workaround so SEB Server is configured as
described in the linked workaround post. I guess I misunderstood the
workaround step-2 to mean uploading the custom SEB config exported from SEB
Server. But, maybe he was referring to a SEB config developed external to
the server.

I have tried uploading both SEB config exports into my Moodle quiz. In my
testing the connection configuration is the only one that opens the correct
Moodle quiz. When I uploaded the exam configuration to Moodle the SEB
config opens to the SEB homepage rather than the Moodle quiz. Maybe this is
the result of the incompatibility with the Moodle deeper integration?

If I ‘hardcode’ the correct start URL into the server’s exam config and
upload that to Moodle, then I would need a unique exam config for every
exam? But, it doesn’t appear that scenario would result in the SEB Server
live monitoring of clients linking to the correct running exam.

**Comment by blanderson22** [comment]
> But, it doesn’t appear that scenario would result in the SEB Server live
> monitoring of clients linking to the correct running exam.

Perhaps here it would’ve been clearer to say, ‘it doesn’t appear that an
exam config .seb appropriately attaches the clients to the live monitoring
of the correct running exam.

**Reply by anhefti** [reply]
The attaching of the SEB clients to the monitoring and to the respective (ad-hoc generated) proctoring rooms is done on the fly by SEB Server sending the connection details to SEB when the SEB connects to SEB Server and the a specific exam. Only the overall proctoring settings are defined within the Exam Configuration. This is an additional service of SEB Server, to create proctoring rooms of defined size on the fly while SEB clients are connecting to SEB Server.

**Comment by blanderson22** [comment]
By the way, currently what I’m testing is, I suppose, my own workaround.

I set my Moodle quiz to require SEB but set the config setting to use the
client config (since the connection config won’t upload to Moodle as a
.seb). But, then I was going to set the required key setting to be that of
the SEB Server config. Next, I added a link to the SEB Server connection
config for the appropriate running exam into the quiz description with
instructions to download the file in order to begin the exam.

This workaround seems to accurately lock down the quiz to require SEB,
while also connecting the client to the running exam’s live monitoring in
SEB server.
